src/Microsoft.AspNet.OData.Shared/EnableQueryAttribute.cs

```diff
@@ -0,0 +1,783 @@
+                    }
+                }
+
+                return queryable;
+            }
+        }
+
+        /// <summary>
+        /// Get the element type.
+        /// </summary>
+        /// <param name="responseValue">The response value.</param>
+        /// <param name="singleResultCollection">The content as SingleResult.Queryable.</param>
+        /// <param name="actionDescriptor">The action context, i.e. action and controller name.</param>
+        /// <returns></returns>
+        internal static Type GetElementType(
+            object responseValue,
+            IQueryable singleResultCollection,
+            IWebApiActionDescriptor actionDescriptor)
+        {
+            Contract.Assert(responseValue != null);
+
+            IEnumerable enumerable = responseValue as IEnumerable;
+            if (enumerable == null)
+            {
+                if (singleResultCollection == null)
+                {
+                    return responseValue.GetType();
+                }
+
+                enumerable = singleResultCollection as IEnumerable;
+            }
+
+            Type elementClrType = TypeHelper.GetImplementedIEnumerableType(enumerable.GetType());
+            if (elementClrType == null)
+            {
+                // The element type cannot be determined because the type of the content
+                // is not IEnumerable<T> or IQueryable<T>.
+                throw Error.InvalidOperation(
+                    SRResources.FailedToRetrieveTypeToBuildEdmModel,
+                    typeof(EnableQueryAttribute).Name,
+                    actionDescriptor.ActionName,
+                    actionDescriptor.ControllerName,
+                    responseValue.GetType().FullName);
+            }
+
+            return elementClrType;
+        }
+
+        /// <summary>
+        /// Get a single or default value from a collection.
+        /// </summary>
+        /// <param name="queryable">The response value as <see cref="IQueryable"/>.</param>
+        /// <param name="actionDescriptor">The action context, i.e. action and controller name.</param>
+        /// <returns></returns>
+        internal static object SingleOrDefault(
+            IQueryable queryable,
+            IWebApiActionDescriptor actionDescriptor)
+        {
+            var enumerator = queryable.GetEnumerator();
+            try
+            {
+                var result = enumerator.MoveNext() ? enumerator.Current : null;
+
+                if (enumerator.MoveNext())
+                {
+                    throw new InvalidOperationException(Error.Format(
+                        SRResources.SingleResultHasMoreThanOneEntity,
+                        actionDescriptor.ActionName,
+                        actionDescriptor.ControllerName,
+                        "SingleResult"));
+                }
+
+                return result;
+            }
+            finally
+            {
+                // Fix for Issue #2097
```

Comment:
I couldn't find [issue #2097](https://github.com/OData/WebApi/issues/2097) for some reason. In any case, is this code comment being tracked? Wondering if it'll be forgotten and turned into code smell later.

Reply:
I think the comment following explains the behavior well enough, I'll remove this line.